vectoriadb-index/src/main/java/jetbrains/vectoriadb/index/KMeansPlusPlusClusterInitializer.java

```diff
@@ -0,0 +1,144 @@
+                            var end = Math.min(start + assignmentSize, numVectors);
+                            var id = i;
+
+                            futures[i] = executors.submit(() -> {
+                                try (var localTracker = mtProgressTracker.createThreadLocalTracker(id)) {
+                                    var localSize = end - start;
+                                    double sumSquaredDistanceToClosestCentroid = 0f;
+                                    for (long k = 0; k < localSize; k++) {
+                                        long vectorIdx = start + k;
+                                        float distanceToClosestCentroid = distancesToClosestCentroid.getAtIndex(ValueLayout.JAVA_FLOAT, vectorIdx);
+                                        // process the distance to the last initialized centroid
+                                        float distanceToLastCentroid = AbstractQuantizer.symmetricDistance(pqVectors, vectorIdx, pqCentroids, lastCentroidIdx, distanceTable, quantizersCount, codeBaseSize);
+                                        if (distanceToLastCentroid < distanceToClosestCentroid) {
+                                            distanceToClosestCentroid = distanceToLastCentroid;
+                                            distancesToClosestCentroid.setAtIndex(ValueLayout.JAVA_FLOAT, vectorIdx, distanceToClosestCentroid);
+                                        }
+
+                                        sumSquaredDistanceToClosestCentroid += weightedDistance(distanceToClosestCentroid);
+                                        localTracker.progress(k * 100.0 / localSize);
+                                    }
+                                    sumSquaredDistancesToClosestCentroid[id] = sumSquaredDistanceToClosestCentroid;
+                                }
+                            });
+                        }
+
+                        for (var future : futures) {
+                            try {
+                                future.get();
+                            } catch (InterruptedException | ExecutionException e) {
+                                throw new RuntimeException("Error during cluster initialization phase in PQ kmeans clustering.", e);
+                            }
+                        }
+
+                        // reduce
+                        double sumSquaredDistanceToClosestCentroid = 0;
+                        for (var i = 0; i < cores; i++) {
+                            sumSquaredDistanceToClosestCentroid += sumSquaredDistancesToClosestCentroid[i];
+                        }
+
+                        // choose a vector to become a new centroid
+                        var randomSum = rng.nextDouble(sumSquaredDistanceToClosestCentroid);
+                        double runningSum = 0;
+                        long vectorToBecomeCentroidIdx = 0;
+                        while (vectorToBecomeCentroidIdx < numVectors - 1) {
```

Comment:
Hi, Kiriill.
Code looks good now.
But still performance implications left.
This cycle have O(N*k) complexity and too heave for single thread processing.
Please try to apply map-reduce approach on calculation of result of distribution based on weighted probably. Otherwise users will wait too long for this loop to be finished for all centroids.

Reply:
Hi Krill,
Please do not work on this until you finish your current task.
It is much more important.

Reply:
I got it. Thanks for noticing! I will fix it after the dotProduct quantizer task.